File: day4/go/part_1.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func countXmas(grid [][]rune) int {
	word := "XMAS"
	directions := [][2]int{
		{-1, 0}, {1, 0}, {0, -1}, {0, 1},
		{-1, -1}, {-1, 1}, {1, -1}, {1, 1},
	}
	count := 0

	valid := func(r, c, dr, dc int) bool {
		for i, ch := range word {
			nr, nc := r+dr*i, c+dc*i
			if nr < 0 || nr >= len(grid) || nc < 0 || nc >= len(grid[0]) {
				return false
			}
			if grid[nr][nc] != ch {
				return false
			}
		}
		return true
	}

	for r := 0; r < len(grid); r++ {
		for c := 0; c < len(grid[0]); c++ {
			for _, dir := range directions {
				if valid(r, c, dir[0], dir[1]) {
					count++
				}
			}
		}
	}

	return count
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	var grid [][]rune
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		grid = append(grid, []rune(scanner.Text()))
	}

	fmt.Println(countXmas(grid))
}

File: day4/java/Part1.java
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.List;

public class Part1 {
    private static int countXmas(char[][] grid) {
        String word = "XMAS";
        int[][] directions = {
            {-1, 0}, {1, 0}, {0, -1}, {0, 1},
            {-1, -1}, {-1, 1}, {1, -1}, {1, 1}
        };
        int count = 0;

        for (int r = 0; r < grid.length; r++) {
            for (int c = 0; c < grid[0].length; c++) {
                for (int[] dir : directions) {
                    if (valid(grid, r, c, dir[0], dir[1], word)) {
                        count++;
                    }
                }
            }
        }

        return count;
    }

    private static boolean valid(char[][] grid, int r, int c, int dr, int dc, String word) {
        for (int i = 0; i < word.length(); i++) {
            int nr = r + dr * i;
            int nc = c + dc * i;
            if (nr < 0 || nr >= grid.length || nc < 0 || nc >= grid[0].length) {
                return false;
            }
            if (grid[nr][nc] != word.charAt(i)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) throws Exception {
        List<String> lines = Files.readAllLines(Paths.get("input.txt"));
        char[][] grid = new char[lines.size()][];
        for (int i = 0; i < lines.size(); i++) {
            grid[i] = lines.get(i).toCharArray();
        }

        System.out.println(countXmas(grid));
    }
}
